src/Chapter20/Listing20.04.ValueTaskAsyncReturn.cs

```diff
@@ -13,45 +15,30 @@ public static async ValueTask<byte[]> CompressAsync(byte[] buffer)
         {
             return buffer;
         }
-        using MemoryStream memoryStream = new();
-        using System.IO.Compression.GZipStream gZipStream =
-            new(
-                memoryStream, 
-                    System.IO.Compression.CompressionMode.Compress);
-        await gZipStream.WriteAsync(buffer.AsMemory(0, buffer.Length));
+        using MemoryStream outputMemoryStream = new();
+        using (GZipStream gZipStream = new(outputMemoryStream, CompressionMode.Compress))
+        {
+            await gZipStream.WriteAsync(buffer.AsMemory(0, buffer.Length));
+        }
 
-        return memoryStream.ToArray();
+        return outputMemoryStream.ToArray();
     }
     #region EXCLUDE
-
-    public static string UnZip(string value)
+    public static async ValueTask<byte[]> Decompress(byte[] buffer)
     {
-        //Transform string into byte[]
-        byte[] byteArray = new byte[value.Length];
-        int index = 0;
-        foreach (char item in value.ToCharArray())
+        if (buffer.Length == 0)
         {
-            byteArray[index++] = (byte)item;
+            return buffer;
         }
 
-        using MemoryStream memoryStream = new(byteArray);
-        using System.IO.Compression.GZipStream gZipStream = new(memoryStream,
-                System.IO.Compression.CompressionMode.Decompress);
-
-         byteArray = new byte[byteArray.Length];
-
-        //Decompress
-        int bytesRead = gZipStream.Read(byteArray, 0, byteArray.Length);
-
-        //Transform byte[] unzip data to string
-        StringBuilder data = new(bytesRead);
-        //Read the number of bytes GZipStream red and do not a for each bytes in
-        //resultByteArray
-        for (int i = 0; i < bytesRead; i++)
+        using MemoryStream inputStream = new(buffer);
+        using MemoryStream uncompressedStream = new();
+        using (GZipStream gZipStream = new(inputStream, CompressionMode.Decompress))
```

Comment:
```suggestion
        using GZipStream gZipStream = new(inputStream, CompressionMode.Decompress);
```

Reply:
I can't do this because this will cause it to fail because the scoping will be different and dispose too late. It's a fun case that I might write a quick blog post about :)

Reply:
Interesting. I would be curious to understand that. Since it appears that the usage should be done.

Reply:
Happy to look at it without you if you want. Totally possible I'm missing something. Took me a long time to figure out that this caused an actual difference in the result though (the scoping of the using). I think because the pointer is at the end of the stream until it gets flushed? But I need to look closer

Reply:
Going to resolve this conversation and move this continued conversation off GitHub.